File: python-server/app/scripting/engine.py
"""Lua script engine: compilation, caching, and execution."""

import threading
from typing import Any, Dict, List, Optional, Tuple

from loguru import logger
from lupa import LuaRuntime  # type: ignore[import-untyped]

from app.scripting.sandbox import (
    MAX_SCRIPT_SIZE,
    HookFlags,
    create_sandboxed_lua,
    parse_hooks,
)


class LuaScriptError(Exception):
    """Raised when a Lua hook execution fails."""


_CTX_HELPERS = """
function _create_ctx(req, resp, provider, model)
    local ctx = {}
    ctx._request = req
    ctx._response = resp
    ctx._unified = nil
    ctx._provider = provider
    ctx._model = model
    ctx._client_protocol = ""
    ctx._provider_protocol = ""
    ctx._meta = {}

    function ctx:get_request() return self._request end
    function ctx:set_request(v) self._request = v end
    function ctx:get_response() return self._response end
    function ctx:set_response(v) self._response = v end
    function ctx:get_unified() return self._unified end
    function ctx:set_unified(v) self._unified = v end
    function ctx:get_provider() return self._provider end
    function ctx:get_model() return self._model end
    function ctx:get_client_protocol() return self._client_protocol end
    function ctx:get_provider_protocol() return self._provider_protocol end
    function ctx:get_meta(k) return self._meta[k] end
    function ctx:set_meta(k, v) self._meta[k] = v end
    return ctx
end
"""


class _CompiledScript:
    __slots__ = ("lua", "hooks")

    def __init__(self, lua: LuaRuntime, hooks: HookFlags) -> None:
        self.lua = lua
        self.hooks = hooks


class LuaEngine:
    """Manages compiled Lua scripts keyed by provider name."""

    def __init__(self) -> None:
        self._scripts: Dict[str, _CompiledScript] = {}
        self._lock = threading.Lock()

    def reload(self, sources: List[Tuple[str, str]]) -> None:
        """Reload scripts from (provider_name, source) pairs.

        On compilation failure the previous script is preserved.
        Providers absent from *sources* are removed.
        """
        new_scripts: Dict[str, _CompiledScript] = {}
        failed: list[str] = []

        for name, source in sources:
            compiled = self._compile(source)
            if compiled is not None:
                logger.info(f"Lua script compiled for provider: {name}")
                new_scripts[name] = compiled
            else:
                logger.error(
                    f"Failed to compile Lua script for provider: {name}, "
                    "keeping previous version"
                )
                failed.append(name)

        with self._lock:
            for name in failed:
                old = self._scripts.get(name)
                if old is not None:
                    logger.warning(f"Retaining previous Lua script version for: {name}")
                    new_scripts[name] = old
            self._scripts = new_scripts

    def reload_from_providers(self, providers: list[Any]) -> None:
        """Reload scripts from provider objects."""
        sources: List[Tuple[str, str]] = []
        for p in providers:
            name = getattr(p, "name", None) or getattr(p, "provider_key", "")
            script = getattr(p, "lua_script", None)
            if script:
                sources.append((name, script))
        self.reload(sources)

    def has_script(self, provider_name: str) -> bool:
        with self._lock:
            return provider_name in self._scripts

    def has_stream_chunk_hook(self, provider_name: str) -> bool:
        with self._lock:
            s = self._scripts.get(provider_name)
            return s.hooks.on_stream_chunk if s else False

    def has_transform_hooks(self, provider_name: str) -> bool:
        with self._lock:
            s = self._scripts.get(provider_name)
            return s.hooks.has_transform_hooks() if s else False

    # =========================================================================
    # Existing hooks (raw JSON level)
    # =========================================================================

    def call_on_request(
        self,
        provider_name: str,
        request: dict,
        model: str,
    ) -> Optional[dict]:
        """Call on_request hook. Returns modified request or None."""
        with self._lock:
            compiled = self._scripts.get(provider_name)
        if compiled is None or not compiled.hooks.on_request:
            return None
        return self._call_hook(
            compiled, "on_request", request, None, None, provider_name, model
        )

    def call_on_response(
        self,
        provider_name: str,
        response: dict,
        model: str,
    ) -> Optional[dict]:
        """Call on_response hook. Returns modified response or None."""
        with self._lock:
            compiled = self._scripts.get(provider_name)
        if compiled is None or not compiled.hooks.on_response:
            return None
        return self._call_hook(
            compiled, "on_response", None, response, None, provider_name, model
        )

    def call_on_stream_chunk(
        self,
        provider_name: str,
        chunk: dict,
        model: str,
    ) -> Optional[dict]:
        """Call on_stream_chunk hook. Returns modified chunk or None."""
        with self._lock:
            compiled = self._scripts.get(provider_name)
        if compiled is None or not compiled.hooks.on_stream_chunk:
            return None
        return self._call_hook(
            compiled, "on_stream_chunk", None, chunk, None, provider_name, model
        )

    # =========================================================================
    # Protocol transform hooks (UIF level)
    # =========================================================================

    def call_on_transform_request_out(
        self,
        provider_name: str,
        request: dict,
        model: str,
        client_protocol: str = "",
        provider_protocol: str = "",
    ) -> Optional[dict]:
        """Call on_transform_request_out (Client → UIF). Returns UIF dict or None."""
        with self._lock:
            compiled = self._scripts.get(provider_name)
        if compiled is None or not compiled.hooks.on_transform_request_out:
            return None
        return self._call_transform_hook(
            compiled,
            "on_transform_request_out",
            request=request,
            response=None,
            unified=None,
            provider_name=provider_name,
            model=model,
            client_protocol=client_protocol,
            provider_protocol=provider_protocol,
            extract="unified",
        )

    def call_on_transform_request_in(
        self,
        provider_name: str,
        unified: dict,
        model: str,
        client_protocol: str = "",
        provider_protocol: str = "",
    ) -> Optional[dict]:
        """Call on_transform_request_in (UIF → Provider). Returns provider dict or None."""
        with self._lock:
            compiled = self._scripts.get(provider_name)
        if compiled is None or not compiled.hooks.on_transform_request_in:
            return None
        return self._call_transform_hook(
            compiled,
            "on_transform_request_in",
            request=None,
            response=None,
            unified=unified,
            provider_name=provider_name,
            model=model,
            client_protocol=client_protocol,
            provider_protocol=provider_protocol,
            extract="request",
        )

    def call_on_transform_response_in(
        self,
        provider_name: str,
        response: dict,
        model: str,
        client_protocol: str = "",
        provider_protocol: str = "",
    ) -> Optional[dict]:
        """Call on_transform_response_in (Provider → UIF). Returns UIF dict or None."""
        with self._lock:
            compiled = self._scripts.get(provider_name)
        if compiled is None or not compiled.hooks.on_transform_response_in:
            return None
        return self._call_transform_hook(
            compiled,
            "on_transform_response_in",
            request=None,
            response=response,
            unified=None,
            provider_name=provider_name,
            model=model,
            client_protocol=client_protocol,
            provider_protocol=provider_protocol,
            extract="unified",
        )

    def call_on_transform_response_out(
        self,
        provider_name: str,
        unified: dict,
        model: str,
        client_protocol: str = "",
        provider_protocol: str = "",
    ) -> Optional[dict]:
        """Call on_transform_response_out (UIF → Client). Returns client dict or None."""
        with self._lock:
            compiled = self._scripts.get(provider_name)
        if compiled is None or not compiled.hooks.on_transform_response_out:
            return None
        return self._call_transform_hook(
            compiled,
            "on_transform_response_out",
            request=None,
            response=None,
            unified=unified,
            provider_name=provider_name,
            model=model,
            client_protocol=client_protocol,
            provider_protocol=provider_protocol,
            extract="response",
        )

    # =========================================================================
    # Internal
    # =========================================================================

    @staticmethod
    def _compile(source: str) -> Optional[_CompiledScript]:
        if len(source) > MAX_SCRIPT_SIZE:
            logger.error(
                f"Lua script too large ({len(source)} bytes, max {MAX_SCRIPT_SIZE})"
            )
            return None

        try:
            lua = create_sandboxed_lua()
            lua.execute(_CTX_HELPERS)
            lua.execute(source)
        except Exception as e:
            logger.error(f"Lua compilation error: {e}")
            return None

        hooks = parse_hooks(lua)
        return _CompiledScript(lua=lua, hooks=hooks)

    @staticmethod
    def _call_hook(
        compiled: _CompiledScript,
        hook_name: str,
        request: Optional[dict],
        response: Optional[dict],
        unified: Optional[dict],
        provider_name: str,
        model: str,
    ) -> Optional[dict]:
        lua = compiled.lua
        g = lua.globals()
        func = g[hook_name]
        if func is None:
            return None

        try:
            req_table = _dict_to_lua_table(lua, request) if request else None
            resp_table = _dict_to_lua_table(lua, response) if response else None

            create_ctx = g["_create_ctx"]
            ctx = create_ctx(req_table, resp_table, provider_name, model)

            func(ctx)

            if hook_name == "on_request":
                result = ctx["_request"]
            else:
                result = ctx["_response"]

            if result is None:
                return None

            return _lua_table_to_dict(result)

        except Exception as e:
            raise LuaScriptError(
                f"Lua {hook_name} error for {provider_name}: {e}"
            ) from e

    @staticmethod
    def _call_transform_hook(
        compiled: _CompiledScript,
        hook_name: str,
        request: Optional[dict],
        response: Optional[dict],
        unified: Optional[dict],
        provider_name: str,
        model: str,
        client_protocol: str,
        provider_protocol: str,
        extract: str,
    ) -> Optional[dict]:
        """Execute a protocol transform hook.

        *extract* determines which ctx field to read after the hook:
        "unified", "request", or "response".
        """
        lua = compiled.lua
        g = lua.globals()
        func = g[hook_name]
        if func is None:
            return None

        try:
            req_table = _dict_to_lua_table(lua, request) if request else None
            resp_table = _dict_to_lua_table(lua, response) if response else None

            create_ctx = g["_create_ctx"]
            ctx = create_ctx(req_table, resp_table, provider_name, model)

            # Set extra fields not covered by _create_ctx
            if unified is not None:
                ctx["_unified"] = _dict_to_lua_table(lua, unified)
            ctx["_client_protocol"] = client_protocol
            ctx["_provider_protocol"] = provider_protocol

            func(ctx)

            field_map = {
                "unified": "_unified",
                "request": "_request",
                "response": "_response",
            }
            result = ctx[field_map[extract]]

            if result is None:
                return None

            return _lua_table_to_dict(result)

        except Exception as e:
            raise LuaScriptError(
                f"Lua {hook_name} error for {provider_name}: {e}"
            ) from e


def _dict_to_lua_table(lua: LuaRuntime, d: dict) -> Any:
    """Recursively convert a Python dict to a Lua table."""
    return _deep_to_lua(lua, d)


def _deep_to_lua(lua: LuaRuntime, obj: Any) -> Any:
    """Deeply convert a Python object to nested Lua tables.

    lupa's table_from only converts one level, so we must recurse and
    call table_from at every nesting level to produce real Lua tables.
    """
    if isinstance(obj, dict):
        return lua.table_from({k: _deep_to_lua(lua, v) for k, v in obj.items()})
    if isinstance(obj, (list, tuple)):
        return lua.table_from({i + 1: _deep_to_lua(lua, v) for i, v in enumerate(obj)})
    return obj


def _lua_table_to_dict(lua_table: Any) -> Any:
    """Recursively convert a lupa Lua table to a Python dict/list."""
    if not hasattr(lua_table, "keys"):
        return lua_table

    keys = list(lua_table.keys())
    if not keys:
        return {}

    # Check if array-like (consecutive integer keys starting from 1)
    if all(isinstance(k, (int, float)) for k in keys):
        int_keys = sorted(int(k) for k in keys)
        if int_keys == list(range(1, len(int_keys) + 1)):
            return [_lua_table_to_dict(lua_table[k]) for k in int_keys]

    return {
        str(k) if not isinstance(k, str) else k: _lua_table_to_dict(lua_table[k])
        for k in keys
    }


_engine: Optional[LuaEngine] = None
_engine_lock = threading.Lock()


def get_lua_engine() -> LuaEngine:
    """Get or create the global LuaEngine singleton."""
    global _engine
    if _engine is None:
        with _engine_lock:
            if _engine is None:
                _engine = LuaEngine()
    return _engine

File: rust-server/src/scripting/engine.rs
//! Lua script engine: compilation, caching, and execution.

use std::collections::HashMap;
use std::sync::atomic::{AtomicU32, Ordering};
use std::sync::{Arc, RwLock};

use mlua::{Function, Lua};
use tracing;

use super::bindings::LuaTransformContext;
use super::sandbox::{create_sandboxed_lua, parse_hooks, HookFlags, MAX_SCRIPT_SIZE};

/// Cached compiled Lua state for a single provider.
struct CompiledScript {
    lua: Lua,
    /// Reset to 0 before each hook call to give a fresh instruction budget.
    instruction_counter: Arc<AtomicU32>,
    hooks: HookFlags,
}

/// Manages compiled Lua scripts keyed by provider name.
pub struct LuaEngine {
    scripts: RwLock<HashMap<String, CompiledScript>>,
}

impl LuaEngine {
    pub fn new() -> Self {
        Self {
            scripts: RwLock::new(HashMap::new()),
        }
    }

    /// Reload scripts from a list of (provider_name, lua_script_source) pairs.
    ///
    /// On compilation failure the previous script for that provider is
    /// preserved. Providers absent from `sources` are removed.
    pub fn reload(&self, sources: Vec<(String, String)>) {
        let mut new_scripts = HashMap::new();
        let mut failed: Vec<String> = Vec::new();

        for (name, source) in &sources {
            match Self::compile(source) {
                Ok(compiled) => {
                    tracing::info!(provider = %name, "Lua script compiled");
                    new_scripts.insert(name.clone(), compiled);
                }
                Err(e) => {
                    tracing::error!(provider = %name, error = %e, "Failed to compile Lua script, keeping previous version");
                    failed.push(name.clone());
                }
            }
        }

        let mut scripts = self.scripts.write().unwrap();

        // Preserve old compiled scripts for providers that failed to compile
        for name in failed {
            if let Some(old) = scripts.remove(&name) {
                tracing::warn!(provider = %name, "Retaining previous Lua script version");
                new_scripts.insert(name, old);
            }
        }

        *scripts = new_scripts;
    }

    /// Check if a provider has a compiled script.
    pub fn has_script(&self, provider_name: &str) -> bool {
        self.scripts.read().unwrap().contains_key(provider_name)
    }

    /// Check if a provider's script defines the `on_stream_chunk` hook.
    pub fn has_stream_chunk_hook(&self, provider_name: &str) -> bool {
        self.scripts
            .read()
            .unwrap()
            .get(provider_name)
            .map(|s| s.hooks.on_stream_chunk)
            .unwrap_or(false)
    }

    /// Check if a provider's script defines any protocol transform hooks.
    pub fn has_transform_hooks(&self, provider_name: &str) -> bool {
        self.scripts
            .read()
            .unwrap()
            .get(provider_name)
            .map(|s| s.hooks.has_transform_hooks())
            .unwrap_or(false)
    }

    /// Call `on_request` hook for the given provider.
    pub fn call_on_request(
        &self,
        provider_name: &str,
        request: serde_json::Value,
        model: &str,
    ) -> Result<Option<serde_json::Value>, String> {
        let scripts = self.scripts.read().unwrap();
        let compiled = match scripts.get(provider_name) {
            Some(c) if c.hooks.on_request => c,
            _ => return Ok(None),
        };

        compiled.instruction_counter.store(0, Ordering::Relaxed);
        let ctx = LuaTransformContext::for_request(request, provider_name, model);
        Self::call_hook(&compiled.lua, "on_request", ctx)
    }

    /// Call `on_response` hook for the given provider.
    pub fn call_on_response(
        &self,
        provider_name: &str,
        response: serde_json::Value,
        model: &str,
    ) -> Result<Option<serde_json::Value>, String> {
        let scripts = self.scripts.read().unwrap();
        let compiled = match scripts.get(provider_name) {
            Some(c) if c.hooks.on_response => c,
            _ => return Ok(None),
        };

        compiled.instruction_counter.store(0, Ordering::Relaxed);
        let ctx = LuaTransformContext::for_response(response, provider_name, model);
        Self::call_hook_response(&compiled.lua, "on_response", ctx)
    }

    /// Call `on_stream_chunk` hook for the given provider.
    pub fn call_on_stream_chunk(
        &self,
        provider_name: &str,
        chunk: serde_json::Value,
        model: &str,
    ) -> Result<Option<serde_json::Value>, String> {
        let scripts = self.scripts.read().unwrap();
        let compiled = match scripts.get(provider_name) {
            Some(c) if c.hooks.on_stream_chunk => c,
            _ => return Ok(None),
        };

        compiled.instruction_counter.store(0, Ordering::Relaxed);
        let ctx = LuaTransformContext::for_response(chunk, provider_name, model);
        Self::call_hook_response(&compiled.lua, "on_stream_chunk", ctx)
    }

    // =========================================================================
    // Protocol Transform Hooks
    // =========================================================================

    /// Call `on_transform_request_out` hook (Client raw JSON → UIF).
    /// Returns the UIF JSON from `ctx.unified` if the hook set it.
    pub fn call_on_transform_request_out(
        &self,
        provider_name: &str,
        request: serde_json::Value,
        model: &str,
        client_protocol: &str,
        provider_protocol: &str,
    ) -> Result<Option<serde_json::Value>, String> {
        let scripts = self.scripts.read().unwrap();
        let compiled = match scripts.get(provider_name) {
            Some(c) if c.hooks.on_transform_request_out => c,
            _ => return Ok(None),
        };

        compiled.instruction_counter.store(0, Ordering::Relaxed);
        let ctx = LuaTransformContext::for_transform(
            Some(request),
            None,
            None,
            provider_name,
            model,
            client_protocol,
            provider_protocol,
        );
        Self::call_hook_unified(&compiled.lua, "on_transform_request_out", ctx)
    }

    /// Call `on_transform_request_in` hook (UIF → Provider JSON).
    /// Returns the provider JSON from `ctx.request` if the hook set it.
    pub fn call_on_transform_request_in(
        &self,
        provider_name: &str,
        unified: serde_json::Value,
        model: &str,
        client_protocol: &str,
        provider_protocol: &str,
    ) -> Result<Option<serde_json::Value>, String> {
        let scripts = self.scripts.read().unwrap();
        let compiled = match scripts.get(provider_name) {
            Some(c) if c.hooks.on_transform_request_in => c,
            _ => return Ok(None),
        };

        compiled.instruction_counter.store(0, Ordering::Relaxed);
        let ctx = LuaTransformContext::for_transform(
            None,
            None,
            Some(unified),
            provider_name,
            model,
            client_protocol,
            provider_protocol,
        );
        Self::call_hook(&compiled.lua, "on_transform_request_in", ctx)
    }

    /// Call `on_transform_response_in` hook (Provider raw JSON → UIF).
    /// Returns the UIF JSON from `ctx.unified` if the hook set it.
    pub fn call_on_transform_response_in(
        &self,
        provider_name: &str,
        response: serde_json::Value,
        model: &str,
        client_protocol: &str,
        provider_protocol: &str,
    ) -> Result<Option<serde_json::Value>, String> {
        let scripts = self.scripts.read().unwrap();
        let compiled = match scripts.get(provider_name) {
            Some(c) if c.hooks.on_transform_response_in => c,
            _ => return Ok(None),
        };

        compiled.instruction_counter.store(0, Ordering::Relaxed);
        let ctx = LuaTransformContext::for_transform(
            None,
            Some(response),
            None,
            provider_name,
            model,
            client_protocol,
            provider_protocol,
        );
        Self::call_hook_unified(&compiled.lua, "on_transform_response_in", ctx)
    }

    /// Call `on_transform_response_out` hook (UIF → Client JSON).
    /// Returns the client JSON from `ctx.response` if the hook set it.
    pub fn call_on_transform_response_out(
        &self,
        provider_name: &str,
        unified: serde_json::Value,
        model: &str,
        client_protocol: &str,
        provider_protocol: &str,
    ) -> Result<Option<serde_json::Value>, String> {
        let scripts = self.scripts.read().unwrap();
        let compiled = match scripts.get(provider_name) {
            Some(c) if c.hooks.on_transform_response_out => c,
            _ => return Ok(None),
        };

        compiled.instruction_counter.store(0, Ordering::Relaxed);
        let ctx = LuaTransformContext::for_transform(
            None,
            None,
            Some(unified),
            provider_name,
            model,
            client_protocol,
            provider_protocol,
        );
        Self::call_hook_response(&compiled.lua, "on_transform_response_out", ctx)
    }

    fn compile(source: &str) -> Result<CompiledScript, String> {
        if source.len() > MAX_SCRIPT_SIZE {
            return Err(format!(
                "Script size ({} bytes) exceeds maximum ({MAX_SCRIPT_SIZE} bytes)",
                source.len(),
            ));
        }

        let (lua, counter) =
            create_sandboxed_lua().map_err(|e| format!("Failed to create Lua runtime: {e}"))?;

        lua.load(source)
            .exec()
            .map_err(|e| format!("Lua compilation error: {e}"))?;

        // Reset counter after compilation (compilation itself consumes instructions)
        counter.store(0, Ordering::Relaxed);

        let hooks = parse_hooks(&lua);

        Ok(CompiledScript {
            lua,
            instruction_counter: counter,
            hooks,
        })
    }

    fn call_hook(
        lua: &Lua,
        hook_name: &str,
        ctx: LuaTransformContext,
    ) -> Result<Option<serde_json::Value>, String> {
        let func: Function = lua
            .globals()
            .get(hook_name)
            .map_err(|e| format!("Failed to get {hook_name}: {e}"))?;

        let ctx = lua
            .create_userdata(ctx)
            .map_err(|e| format!("Failed to create context: {e}"))?;

        func.call::<()>(ctx.clone())
            .map_err(|e| format!("Lua {hook_name} error: {e}"))?;

        let ctx_ref = ctx
            .borrow::<LuaTransformContext>()
            .map_err(|e| format!("Failed to borrow context: {e}"))?;

        Ok(ctx_ref.request.clone())
    }

    fn call_hook_response(
        lua: &Lua,
        hook_name: &str,
        ctx: LuaTransformContext,
    ) -> Result<Option<serde_json::Value>, String> {
        let func: Function = lua
            .globals()
            .get(hook_name)
            .map_err(|e| format!("Failed to get {hook_name}: {e}"))?;

        let ctx = lua
            .create_userdata(ctx)
            .map_err(|e| format!("Failed to create context: {e}"))?;

        func.call::<()>(ctx.clone())
            .map_err(|e| format!("Lua {hook_name} error: {e}"))?;

        let ctx_ref = ctx
            .borrow::<LuaTransformContext>()
            .map_err(|e| format!("Failed to borrow context: {e}"))?;

        Ok(ctx_ref.response.clone())
    }

    /// Execute a hook and extract `ctx.unified` (for transform hooks that produce UIF).
    fn call_hook_unified(
        lua: &Lua,
        hook_name: &str,
        ctx: LuaTransformContext,
    ) -> Result<Option<serde_json::Value>, String> {
        let func: Function = lua
            .globals()
            .get(hook_name)
            .map_err(|e| format!("Failed to get {hook_name}: {e}"))?;

        let ctx = lua
            .create_userdata(ctx)
            .map_err(|e| format!("Failed to create context: {e}"))?;

        func.call::<()>(ctx.clone())
            .map_err(|e| format!("Lua {hook_name} error: {e}"))?;

        let ctx_ref = ctx
            .borrow::<LuaTransformContext>()
            .map_err(|e| format!("Failed to borrow context: {e}"))?;

        Ok(ctx_ref.unified.clone())
    }
}

impl Default for LuaEngine {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_engine_new() {
        let engine = LuaEngine::new();
        assert!(!engine.has_script("test"));
    }

    #[test]
    fn test_engine_reload() {
        let engine = LuaEngine::new();
        engine.reload(vec![(
            "provider-a".to_string(),
            "function on_request(ctx) end".to_string(),
        )]);
        assert!(engine.has_script("provider-a"));
        assert!(!engine.has_script("provider-b"));
    }

    #[test]
    fn test_engine_call_on_request() {
        let engine = LuaEngine::new();
        engine.reload(vec![(
            "provider-a".to_string(),
            r#"
            function on_request(ctx)
                local req = ctx:get_request()
                req.temperature = 0.5
                ctx:set_request(req)
            end
            "#
            .to_string(),
        )]);

        let request = serde_json::json!({"model": "gpt-4", "temperature": 1.0});
        let result = engine
            .call_on_request("provider-a", request, "gpt-4")
            .unwrap();

        assert!(result.is_some());
        let modified = result.unwrap();
        assert_eq!(modified["temperature"], 0.5);
    }

    #[test]
    fn test_engine_call_on_response() {
        let engine = LuaEngine::new();
        engine.reload(vec![(
            "provider-a".to_string(),
            r#"
            function on_response(ctx)
                local resp = ctx:get_response()
                resp.custom_field = "added"
                ctx:set_response(resp)
            end
            "#
            .to_string(),
        )]);

        let response = serde_json::json!({"model": "gpt-4"});
        let result = engine
            .call_on_response("provider-a", response, "gpt-4")
            .unwrap();

        assert!(result.is_some());
        let modified = result.unwrap();
        assert_eq!(modified["custom_field"], "added");
    }

    #[test]
    fn test_engine_no_script() {
        let engine = LuaEngine::new();
        let request = serde_json::json!({"model": "gpt-4"});
        let result = engine
            .call_on_request("nonexistent", request, "gpt-4")
            .unwrap();
        assert!(result.is_none());
    }

    #[test]
    fn test_engine_invalid_script() {
        let engine = LuaEngine::new();
        engine.reload(vec![(
            "bad".to_string(),
            "this is not valid lua {{{{".to_string(),
        )]);
        assert!(!engine.has_script("bad"));
    }

    #[test]
    fn test_engine_rollback_on_failure() {
        let engine = LuaEngine::new();

        // Load a valid script first
        engine.reload(vec![(
            "provider-a".to_string(),
            "function on_request(ctx) end".to_string(),
        )]);
        assert!(engine.has_script("provider-a"));

        // Reload with an invalid script for provider-a — old version should be kept
        engine.reload(vec![(
            "provider-a".to_string(),
            "invalid lua {{{{".to_string(),
        )]);
        assert!(
            engine.has_script("provider-a"),
            "Old script should be preserved on compilation failure"
        );
    }

    #[test]
    fn test_engine_instruction_limit() {
        let engine = LuaEngine::new();
        engine.reload(vec![(
            "loop".to_string(),
            "function on_request(ctx) while true do end end".to_string(),
        )]);

        let request = serde_json::json!({"model": "test"});
        let result = engine.call_on_request("loop", request, "test");
        assert!(result.is_err());
        let err = result.unwrap_err();
        assert!(
            err.contains("instruction limit"),
            "Expected instruction limit error, got: {err}"
        );
    }

    // -------------------------------------------------------------------------
    // Protocol Transform Hook Tests
    // -------------------------------------------------------------------------

    #[test]
    fn test_transform_request_out() {
        let engine = LuaEngine::new();
        engine.reload(vec![(
            "provider-a".to_string(),
            r#"
            function on_transform_request_out(ctx)
                local raw = ctx:get_request()
                ctx:set_unified({
                    model = raw.model,
                    messages = {},
                    parameters = { stream = false }
                })
            end
            "#
            .to_string(),
        )]);

        assert!(engine.has_transform_hooks("provider-a"));

        let request = serde_json::json!({"model": "gpt-4", "messages": []});
        let result = engine
            .call_on_transform_request_out("provider-a", request, "gpt-4", "openai", "anthropic")
            .unwrap();

        assert!(result.is_some());
        let uif = result.unwrap();
        assert_eq!(uif["model"], "gpt-4");
        assert!(uif["parameters"]["stream"].is_boolean());
    }

    #[test]
    fn test_transform_request_in() {
        let engine = LuaEngine::new();
        engine.reload(vec![(
            "provider-a".to_string(),
            r#"
            function on_transform_request_in(ctx)
                local uif = ctx:get_unified()
                ctx:set_request({
                    model = uif.model,
                    prompt = "converted"
                })
            end
            "#
            .to_string(),
        )]);

        let unified = serde_json::json!({"model": "gpt-4", "messages": []});
        let result = engine
            .call_on_transform_request_in("provider-a", unified, "gpt-4", "openai", "anthropic")
            .unwrap();

        assert!(result.is_some());
        let provider_req = result.unwrap();
        assert_eq!(provider_req["prompt"], "converted");
    }

    #[test]
    fn test_transform_response_in() {
        let engine = LuaEngine::new();
        engine.reload(vec![(
            "provider-a".to_string(),
            r#"
            function on_transform_response_in(ctx)
                local resp = ctx:get_response()
                ctx:set_unified({
                    id = "msg_1",
                    model = resp.model,
                    content = {{ type = "text", text = "hello" }}
                })
            end
            "#
            .to_string(),
        )]);

        let response = serde_json::json!({"model": "gpt-4", "choices": []});
        let result = engine
            .call_on_transform_response_in("provider-a", response, "gpt-4", "openai", "anthropic")
            .unwrap();

        assert!(result.is_some());
        let uif = result.unwrap();
        assert_eq!(uif["id"], "msg_1");
    }

    #[test]
    fn test_transform_response_out() {
        let engine = LuaEngine::new();
        engine.reload(vec![(
            "provider-a".to_string(),
            r#"
            function on_transform_response_out(ctx)
                local uif = ctx:get_unified()
                ctx:set_response({
                    id = uif.id,
                    object = "chat.completion",
                    model = uif.model
                })
            end
            "#
            .to_string(),
        )]);

        let unified = serde_json::json!({"id": "msg_1", "model": "gpt-4", "content": []});
        let result = engine
            .call_on_transform_response_out("provider-a", unified, "gpt-4", "openai", "anthropic")
            .unwrap();

        assert!(result.is_some());
        let client_resp = result.unwrap();
        assert_eq!(client_resp["object"], "chat.completion");
    }

    #[test]
    fn test_no_transform_hooks_returns_none() {
        let engine = LuaEngine::new();
        engine.reload(vec![(
            "provider-a".to_string(),
            "function on_request(ctx) end".to_string(),
        )]);

        assert!(!engine.has_transform_hooks("provider-a"));

        let request = serde_json::json!({"model": "gpt-4"});
        let result = engine
            .call_on_transform_request_out("provider-a", request, "gpt-4", "openai", "anthropic")
            .unwrap();
        assert!(result.is_none());
    }

    #[test]
    fn test_transform_hook_has_protocol_context() {
        let engine = LuaEngine::new();
        engine.reload(vec![(
            "provider-a".to_string(),
            r#"
            function on_transform_request_out(ctx)
                local cp = ctx:get_client_protocol()
                local pp = ctx:get_provider_protocol()
                ctx:set_unified({
                    model = "test",
                    client = cp,
                    provider = pp
                })
            end
            "#
            .to_string(),
        )]);

        let request = serde_json::json!({"model": "test"});
        let result = engine
            .call_on_transform_request_out("provider-a", request, "test", "openai", "anthropic")
            .unwrap()
            .unwrap();

        assert_eq!(result["client"], "openai");
        assert_eq!(result["provider"], "anthropic");
    }
}
